known Customer Credential-Wallet (#1346)
TBD54566975/developer.tbd.website: site/docs/tbdex/issuer/kcc-wallet.mdx

---
sidebar_position: 2
title: KCC Wallet Flow
hide_title: true
---

<LanguageSwitcher languages="JavaScript, Kotlin" />

# Requesting a Known Customer Credential

> Known Customer Credentials (KCCs) are [Verifiable Credentials](/docs/web5/learn/verifiable-credentials) designed to streamline the [Know Your Customer (KYC)](/) process for tbDEX protocol users. KCCs help in gaining access to PFIs providing regulated financial services.

Reviewer: ```suggestion
Known Customer Credentials (KCCs) are [Verifiable Credentials](/docs/web5/learn/verifiable-credentials) designed to streamline the [Know Your Customer (KYC)](/) process for tbDEX protocol users. KCCs help in gaining access to PFIs that provide regulated financial services.
```

For a comprehensive exploration of the practical applications and compliance considerations of KCCs, refer to our [KCC Compliance Guide](/). It provides detailed examples and insights into how KCCs align with certain regulatory requirements. This offers valuable context for those overseeing the implementation and management of KCCs.

In this guide, we'll cover:

- **Identity Verification (IDV)**: How your Wallet App initiates and supports the IDV process conformant with [Self Issued OpenID Provider V2 (SIOPv2)](https://openid.net/specs/openid-connect-self-issued-v2-1_0.html) and [OpenID for Verifiable Presentations (OID4VP)](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html), ensuring users can securely verify their identity.

> - **Credential Request**: The steps needed for your Wallet App to request a KCC **post** successful IDV, on a users behalf.

Reviewer: ```suggestion
- **Credential Request**: The steps needed for your Wallet App to request a KCC on a user's behalf after successful IDV.
```

> - **Credential Management**: How the Wallet utilizes existing credentials for KCC requests and secure storage of the KCC itself.

Reviewer: ```suggestion
- **Credential Management**: How the Wallet utilizes existing credentials for KCC requests and securely stores the KCC itself.
```

:::info

> For a detailed guide on Known Customer Credentials for **Issuers**, please refer to our [Issuer's Known Customer Credential Guide](/)

Reviewer: ```suggestion
For a detailed guide on Known Customer Credentials for **Issuers**, please refer to the [Issuer's Known Customer Credential Guide](/)
```

      language: 'JavaScript',
      title: 'wallet.js',
    },
    {
      snippetName: 'knownCustomerCredentialHandleIssuerResponseWalletKT',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

:::info
The `credential_issuer` URL is crucial for the next steps, as it will be used to construct the URLs for fetching the Issuer's Metadata and the Issuer's Authorization Server Metadata.
:::

2. Fetch the Issuer's Metadata from the `credential_issuers` URL

To construct the URL to fetch the Issuer's metadata, the Wallet needs to append `/.well-known/openid-credential-issuer` to the `credential_issuer` URL provided in the `credential_offer`:

<Shnip
  snippets={[
    {
      snippetName: 'knownCustomerCredentialFetchIssuerMetadataWalletJS',
      language: 'JavaScript',
      title: 'wallet.js',
    },
    {
      snippetName: 'knownCustomerCredentialFetchIssuerMetadataWalletKT',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

The Issuer's Metadata includes information about the Issuer's capabilities, credential endpoint, supported credential configurations, and more. Some of the key information included is:

> - `credential_issuer`: The base URL address of the Issuer's service. It is used by the Wallet to retrieve the Issuer's metadata and the Issuer's Authorization Server Metadata.

Reviewer: ```suggestion
- `credential_issuer`: The base URL address of the Issuer's service. It is used by the Wallet to retrieve the Issuer's Metadata and the Issuer's Authorization Server Metadata.
```

  "credential_endpoint": "https://issuer.example.com/credentials",
  "credential_configurations_supported": {
    "KnownCustomerCredential": {
      "format": "jwt_vc_json",
      "scope": "CustomerIdentity",
      "cryptographic_binding_methods_supported": ["did:example"],
      "credential_signing_alg_values_supported": ["ES256"],
      "credential_definition": {
        "type": ["VerifiableCredential", "KnownCustomerCredential"],
        "credentialSubject": {
          "country": {
            "display": [
              {
                "name": "Country",
                "locale": "en-US"
              }
            ]
          }
        }
      },
      "proof_types_supported": {
        "jwt": {
          "proof_signing_alg_values_supported": ["ES256"]
        }
      },
      "display": [
        {
          "name": "Known Customer Credential",
          "locale": "en-US",
          "logo": {
            "url": "https://issuer.example.com/public/logo.png",
            "alt_text": "Issuer Logo"
          },
          "background_color": "#FFFFFF",
          "text_color": "#000000"
        }
      ]
    }
  }
}

```

</details>

3. Fetch the Authorization Server's Metadata

> Before the Wallet can request a KCC from the Issuer. In the same `handleIssuerResponse()` the Wallet needs to also fetch the Issuer's Authorization Server Metadata. Similarly, by appending `/.well-known/oauth-authorization-server` to the `credential_issuer` URL:

Reviewer: ```suggestion
Before the Wallet can request a KCC from the Issuer, the Wallet needs to also fetch the Issuer's Authorization Server Metadata by appending `/.well-known/oauth-authorization-server` to the `credential_issuer` URL in `handleIssuerResponse()`:
```

      language: 'JavaScript',
      title: 'wallet.js',
    },
    {
      snippetName: 'knownCustomerCredentialFetchAuthServerMetadataWalletKT',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

The Server Authorization Metadata can look like this:

<details open>
<summary>Server Authorization Metadata</summary>

```javascript
// Issuer.js

{
   "issuer": "https://issuer.example.com", // URL of the Credential Issuer
   "token_endpoint": "https://issuer.example.com/token", // URL for the Access Token Request
};
```

</details>

- `issuer`: The base URL of the Issuer's service.
- `token_endpoint`: The endpoint where the Wallet can exchange the `pre_authorization_code` receieved in the `credential_offer` for an access token.

4. Fetch the Access Token from the Issuer's `/token` endpoint:

<Shnip
  snippets={[
    {
      snippetName: 'knownCustomerCredentialFetchAccessTokenWalletJS',
      language: 'JavaScript',
      title: 'wallet.js',
    },
    {
      snippetName: 'knownCustomerCredentialFetchAccessTokenWalletKT',
      language: 'Kotlin',
      title: 'Application.kt',
    },
  ]}
/>

:::info

> If the IDV process is incomplete, the Issuer will return an `authorization_pending` error. In this case, the Wallet should wait for a short period before attempting another request to the Issuer's `/token` endpoint. This will allow time for the IDV process to be completed. The Wallet should continue these periodic attempts until an `access_token` is successfully received, indicating that the IDV process has concluded and authorization has been granted.

Reviewer: should we show the code in a loop with a little sleep? something like we do when [polling tbdex messages from PFIs](https://deploy-preview-1346--tbd-website-developer.netlify.app/docs/tbdex/wallet/receive-quote#polling-for-quote)

      title: 'Application.kt',
    },
  ]}
/>

:::info
To learn more about the **fields** within the `payload` and `header`, please refer to [OpenID's specs](https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#section-7.2.1.1-2.1.1).

The Known Customer Credential will be a signed VC [JSON Web Token](https://jwt.io/) for secure transmission.
:::

To convert the signed VC JWT into a `VerifiableCredential` object you can use the [`parseJwt()` method](/docs/web5/build/verifiable-credentials/jwt-to-vc#decoding-jwt). Here is what a decoded **Known Customer Credential** would look like:

```javascript
{
  "vcDataModel": {
    "@context": ["https://www.w3.org/2018/credentials/v1"],
    "type": ["VerifiableCredential", "KnownCustomerCredential"],
    "id": "urn:uuid:12345678-1234-5678-1234-567812345678",
    "issuer": "did:dht:z6Mkn4w6nSaWe4fjNJRvaHZwFnMm5VexvjzDeozEu2G7jC34", // Issuer's DID
    "issuanceDate": "2024-01-01T19:23:24Z",
    "expirationDate": "2026-05-19T08:02:04Z"
    "credentialSubject": {
      "id": "did:dht:z6MkjGSeekPGE9QfczHWyW8v2ZzJU68kqSHzV7L2dmQyuyDu", // User's DID
      "country_of_residence": "US"
    }
  }
}
````

> To the user an example KCC may look like this:

Reviewer: ```suggestion
To the user, an example KCC may look like this:
```

      title: 'Application.kt',
    },
  ]}
/>

The Issuer's Metadata includes information about the Issuer's capabilities, credential endpoint, supported credential configurations, and more. Some of the key information included is:

- `credential_issuer`: The base URL address of the Issuer's service. It is used by the Wallet to retrieve the Issuer's Metadata and the Issuer's Authorization Server Metadata.
- `credential_endpoint`: The endpoint the Wallet will use to submit requests to issue credentials. After obtaining an access token, the Wallet sends a credential request to this endpoint.
- `credential_configurations_supported`: Defines the format `jwt_vc_json`, and the credential, including informative display markup.

Here is an example of an Issuer's Metadata:

<details close>
<summary>Issuer's Metadata</summary>

```javascript
// Issuer.js

{
  "credential_issuer": "https://issuer.example.com",
  "credential_endpoint": "https://issuer.example.com/credentials",
  "credential_configurations_supported": {
    "KnownCustomerCredential": {
      "format": "jwt_vc_json",
      "scope": "CustomerIdentity",
      "cryptographic_binding_methods_supported": ["did:example"],

>       "credential_signing_alg_values_supported": ["ES256"],

Reviewer 2: ```suggestion
      "credential_signing_alg_values_supported": ["ES256"],
```

believe this should be `EdDSA` and `ES256K`

      "credential_definition": {
        "type": ["VerifiableCredential", "KnownCustomerCredential"],
        "credentialSubject": {
          "country": {
            "display": [
              {
                "name": "Country",
                "locale": "en-US"
              }
            ]
          }
        }
      },
      "proof_types_supported": {
        "jwt": {

>           "proof_signing_alg_values_supported": ["ES256"]

Reviewer 2: same here

    <p>

```javascript
{
  "vcDataModel": {
    "@context": ["https://www.w3.org/2018/credentials/v1"],
    "type": ["VerifiableCredential", "KnownCustomerCredential"],
    "id": "urn:uuid:12345678-1234-5678-1234-567812345678",
    "issuer": "did:dht:z6Mkn4w6nSaWe4fjNJRvaHZwFnMm5VexvjzDeozEu2G7jC34",
    "issuanceDate": "2024-01-01T19:23:24Z",
    "expirationDate": "2026-05-19T08:02:04Z",
    "evidence": [
      {
      kind: "document_verification",
      checks: ["passport", "utility_bill"]
      },
      {
        kind: "sanction_screening",
        checks: ["PEP"]
      }
    ],
    "credentialSubject": {
      "id": "did:dht:z6MkjGSeekPGE9QfczHWyW8v2ZzJU68kqSHzV7L2dmQyuyDu",
      "country_of_residence": "US",
      "tier": "Gold"
    }
  }
}
```

   </p>

   </div>

<div language="Kotlin">
<p>

```javascript
{
  "@context": ["https://www.w3.org/2018/credentials/v1"],
  "type": ["VerifiableCredential", "KnownCustomerCredential"],
  "id": "urn:uuid:12345678-1234-5678-1234-567812345678",
  "issuer": "did:dht:z6Mkn4w6nSaWe4fjNJRvaHZwFnMm5VexvjzDeozEu2G7jC34",
  "issuanceDate": "2024-01-01T19:23:24Z",
  "expirationDate": "2026-05-19T08:02:04Z",
  "evidence": [
    {

>     kind: "document_verification",

Reviewer: ```suggestion
      kind: "document_verification",
```
